### ISSUE TITLE
Sub class validator behaviour loses type.

### ISSUE THREAD

**Issue by dementedhedgehog** [samuelcolvin/pydantic#147]
There's some non-obvious behaviour that loses types in pydantic in the following case.
```
    import pydantic
    
    class IntSub(int):
        pass

    class IntConf(pydantic.BaseModel):
        count: IntSub
    
    c = IntSub(42)
    x = IntConf(count=c)
    assert type(x.count) is IntSub
```
This is because find_validators() checks whether count is a subclass of int and if it is returns a (int, number_size_validator) pair of validators.  So upon creation x.count is set to int(c).

This could be fixed by replacing the int() validator with the following:

```
def new_int_validator(x):
    try:
        int(x)
    except ValueError:
        raise SomeDescriptiveError("blah blah")
    return x
```

This is more strongly typed and since we appear to be caring about typing if we're using pydantic, better .. I believe.  But this is a breaking change since things like IntSub(3.14) and IntSub("2") will stop working.

Same goes for float subtypes .. I assume.  

An alternative is to require a cast constructor for subclasses and use that to check the type instead, e.g. consider what pydantic does without the explicit c=IntSub(42) step above.  So perhaps a better check is that the input is already of the expected type..  if not try a cast constructor, if that fails we tried.

**Comment by samuelcolvin** [comment]
Humm, I see the problem, but I definitely want `3.14` and `"2"` to work with int fields.

It looks like it's not properly documented by the solution is to use `get_validators` like [constr does](https://github.com/samuelcolvin/pydantic/blob/master/pydantic/types.py#L48-L77).

You'll want something like 

```py
class IntSub(int):
    @classmethod
    def get_validators(cls):
        yield cls.myvalidator

    @classmethod
    def myvalidator(cls, v):
        return cls(v)
````

Let me know if that works.

**Comment by dementedhedgehog** [comment]
This does work thanks.

**Comment by binderclip** [comment]
Hi @samuelcolvin , 

I want use a custom data type, and got "**pydantic allows custom data types to be defined**" in the doc but no more info about that,
then found `get_validators` to be the solution in `*(get_validators() if get_validators else find_validators(self.type_))` when dealing with exception log,
finally `get_validators` as keyword to get here.

It will be better if there are something about "custom data types" in the doc. 👍

**Comment by samuelcolvin** [comment]
Absolutely, couldn't agree more. Would you like to submit a PR to add those docs?

**Comment by binderclip** [comment]
Okey! Try to do that in the Weekend.

**Comment by samuelcolvin** [comment]
fix via #157
